sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala

```diff
@@ -183,9 +183,11 @@ case class CatalogTable(
 
   import CatalogTable._
 
-  /** schema of this table's partition columns */
-  def partitionSchema: StructType = StructType(schema.filter {
-    c => partitionColumnNames.contains(c.name)
+  /** schema of this table's partition columns
+   * keep the schema order with partitionColumnNames
+   */
+  def partitionSchema: StructType = StructType(partitionColumnNames.flatMap {
+    p => schema.filter(_.name == p)
```

Comment:
use `partitionColumnNames.map(c => schema.find(_.name == c).getOrElse(throw ...))`, each column name in `partitionColumnNames` must match one and only one field in schema.